### PR TITLE
Handle constant `null`s in pointer pattern matching

### PR DESCRIPTION
**Pull request by jjonescz**
Fixes https://github.com/dotnet/roslyn/issues/70048.
Alternative to https://github.com/dotnet/roslyn/pull/70055.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs [hunk 1]
```diff
@@ -1630,15 +1630,15 @@ void addSubpatternsForTuple(ImmutableArray<TypeWithAnnotations> elementTypes)
 
             BoundExpression value = BindExpressionForPattern(inputType, node.Expression, ref hasErrors, diagnostics, out var constantValueOpt, out _, out Conversion patternConversion);
             ExpressionSyntax innerExpression = SkipParensAndNullSuppressions(node.Expression, diagnostics, ref hasErrors);
-            RoslynDebug.Assert(value.Type is { });
+            var type = value.Type ?? inputType;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:1633]: >var type = value.Type ?? inputType; [](http://example.com/codeflow?start=12&length=35)

```Debug.Assert(type is { });```?

File: src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs [hunk 2]
```diff
@@ -679,10 +679,10 @@ private Tests MakeTestsAndBindingsForNegatedPattern(BoundDagTemp input, BoundNeg
             BoundRelationalPattern rel,
             out BoundDagTemp output)
         {
-            Debug.Assert(rel.Value.Type is not null);
+            var type = rel.Value.Type ?? input.Type;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs:682]: >var type = rel.Value.Type ?? input.Type; [](http://example.com/codeflow?start=12&length=40)

```Debug.Assert(type is not null);```?

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 2)

**Pull request merged**